prov/util/src/util_srx.c

```diff
@@ -0,0 +1,1187 @@
+			fi_addr_t addr, uint64_t size, uint64_t tag)
+{
+	struct util_rx_entry *util_entry;
+
+	util_entry = util_alloc_rx_entry(srx);
+	if (!util_entry)
+		return NULL;
+
+	util_entry->peer_entry.owner_context = NULL;
+	util_entry->peer_entry.size = size;
+	util_entry->peer_entry.addr = addr;
+	util_entry->peer_entry.tag = tag;
+
+	return util_entry;
+}
+
+static bool util_adjust_multi_recv(struct util_srx_ctx *srx,
+		struct fi_peer_rx_entry *rx_entry, size_t len)
+{
+	size_t left;
+	void *new_base;
+
+	left = rx_entry->iov[0].iov_len - len;
+
+	new_base = (void *) ((uintptr_t) rx_entry->iov[0].iov_base + len);
+	rx_entry->iov[0].iov_len = left;
+	rx_entry->iov[0].iov_base = new_base;
+	rx_entry->size = left;
+
+	return left < srx->min_multi_recv_size;
+}
+
+static struct util_rx_entry *util_process_multi_recv(struct util_srx_ctx *srx,
+		struct slist *queue, fi_addr_t addr, size_t size,
+		struct util_rx_entry *owner_entry)
+{
+	struct util_rx_entry *util_entry;
+
+	util_entry = util_get_recv_entry(srx,
+					 owner_entry->peer_entry.iov,
+					 owner_entry->peer_entry.desc,
+					 owner_entry->peer_entry.count, addr,
+					 owner_entry->peer_entry.context,
+					 owner_entry->peer_entry.tag,
+					 owner_entry->ignore,
+					 owner_entry->peer_entry.flags &
+					 (~FI_MULTI_RECV));
+	if (!util_entry)
+		return NULL;
+
+	if (util_adjust_multi_recv(srx, &owner_entry->peer_entry, size))
+		slist_remove_head(queue);
+
+	util_entry->peer_entry.owner_context = owner_entry;
+	owner_entry->multi_recv_ref++;
+
+	return util_entry;
+}
+
+static int util_match_msg(struct fid_peer_srx *srx, fi_addr_t addr, size_t size,
+			  struct fi_peer_rx_entry **rx_entry)
+{
+	struct util_srx_ctx *srx_ctx;
+	struct util_rx_entry *util_entry;
+	struct slist_entry *head;
+	int ret = FI_SUCCESS;
+
+	srx_ctx = srx->ep_fid.fid.context;
+	if (slist_empty(&srx_ctx->msg_queue)) {
+		util_entry = util_init_unexp(srx_ctx, addr, size, 0);
+		if (!util_entry)
+			return -FI_ENOMEM;
+		util_entry->peer_entry.srx = srx;
+		ret = -FI_ENOENT;
+	} else {
+		head = srx_ctx->msg_queue.head;
+		util_entry = container_of(head, struct util_rx_entry,
+					  peer_entry);
+		if (util_entry->peer_entry.flags & FI_MULTI_RECV) {
+			util_entry = util_process_multi_recv(srx_ctx,
+				&srx_ctx->msg_queue, addr, size, util_entry);
+			if (!util_entry) {
+				FI_WARN(&core_prov, FI_LOG_EP_CTRL,
+					"cannot allocate multi receive "
+					"buffer\n");
+				return -FI_ENOMEM;
+			}
+		} else {
+			(void) slist_remove_head(&srx_ctx->msg_queue);
+		}
+		util_entry->peer_entry.srx = srx;
+		srx_ctx->update_func(srx_ctx, util_entry);
+	}
+	*rx_entry = &util_entry->peer_entry;
+	return ret;
+}
+
+static int util_get_msg(struct fid_peer_srx *srx, fi_addr_t addr,
+		        size_t size, struct fi_peer_rx_entry **rx_entry)
+{
+	struct util_srx_ctx *srx_ctx;
+	struct util_rx_entry *util_entry, *any_entry;
+	struct slist *queue;
+
+	srx_ctx = srx->ep_fid.fid.context;
+	assert(ofi_genlock_held(srx_ctx->lock));
```

Comment:
Would you mind pointing to me where shm grab such lock in ep progress?

Reply:
Well I think I need to disable the path you are talking about. That is the path for when the application calls fi_srx_context without FI_PEER (so sharing an srx between shm EPs). I implemented that for completion but I don't think that is supported in shm without some extra domain locking. The "normal" path for shm is where each EP owns the srx so we use the EP lock [here](https://github.com/aingerson/libfabric/blob/peer/prov/shm/src/smr_ep.c#L1338)

Reply:
I see, so it's actually a ep lock. Could be a workaround for efa as well (I mean make efa and shm share the same ep lock). I know it's not recommended though.

Reply:
When shm is working as a peer, it uses the ep lock to manage its internal ep resources but the srx lock (owned by the peer) is in charge of the srx resource locking (which is the one efa needs to grab before calling progress for shm).